LibraryInstrumentationGuidance.md

```diff
@@ -0,0 +1,115 @@
+# Guidance for instrumenting libraries with Diagnostic Source 
+
+This document provides guidance for adding Diagnostic Source instrumentation to external libraries, which will allow Application Insights and other monitoring solutions to collect meaningful and rich telemetry.
+
+## Diagnostic Source and Activities
+
+[Diagnostic Source][DiagnosticSourceGuide] is a simple module that allows code to be instrumented for production-time logging of rich data payloads for consumption within the process that was instrumented. At runtime consumers can dynamically discover data sources and subscribe to the ones of interest.
+
+[Activity][ActivityGuide] is a class that allows storing and accessing diagnostics context and consuming it with logging system.
+
+Both Diagnostic Source and Activity have been used to instrument [System.Net.Http][SystemNetHttp] and [Microsoft.AspNetCore.Hosting][MicrosoftAspNetCoreHosting].
+
+More recently two new libraries were instrumented and that work was the basis for this guidance. These libraries are client SDKs for [Azure Event Hubs][MicrosoftEventHubs] and [Azure Service Bus][MicrosoftServiceBus], both of which support very high throughput scenarios.
+
+[This document][DiagnosticSourceActivityHowto] goes into more details on how to efficiently use Diagnostic Source.
+
+## What should be instrumented
+
+The goal of instrumentation is to give users the visibility to how particular operations are being performed inside the library. This information can be later used to diagnose performance or issues. It is up to the library authors to identify operations that the library performs and are worthy of monitoring. This doesn't have to be limited to instrumenting the exposed API methods but can also cover more specific internal logic (like outgoing service calls, retries, locking, cache utilization, awaiting system events, etc.) that is invoked during processing. This will allow the users to get a good understanding of what's going on under the hood when they need it. 
+
+## Instrumentation 
+
+In the simplest case the operation that is being monitored needs to be wrapped by an activity. However in order to minimize performance impact the activity should be only created if there is any listener waiting for it
+
+```csharp
+    static DiagnosticListener source = new DiagnosticListener("Example.MyLibrary");
+
+    Activity activity = null;
+    // create activity only when requested
+    if (source.IsEnabled() && source.IsEnabled("Example.MyLibrary.MyOperation"))
+    {
+        activity = new Activity("Example.MyLibrary.MyOperation");
```

Comment:
not to overcomplicate this section. Perhaps create a separate one. We have a promise to distributed trace to propagate all incoming headers/context to all outgoing calls. So even if `IsEnabled` returned `false` you may want to create lightweight `Activity` and still send headers. 

We  discussed it with @lmolkova. @lmolkova have you thought about it more?

Reply:
@SergeyKanzhelev we are far from reading any headers except Request-Id. We don't know yet how Activity would look like and what headers to propagate blindly or modify. I suggest we update guidance as soon as we know the details and we'd refer to the propagation standard from here. 

@TomMilos BTW, I just noticed that this guide does not cover injection/extraction at all. I think it worth mentioning.

Reply:
Even with `Request-ID` we may need a feature to propagate ID and context even if telemetry was sampled out. Alternatively we need to say that `IsEnabled` means not participating in distributed tracing at all, not that this Activity was sampled out

Reply:
now IsEnabled serves 2 purposes:
1. preventing unwanted requests to be instrumented (e.g. http calls to AI that report telemetry)
we probably do not want anything to be propagated, but let's not overcomplicate and let's say we don't care if context will be injected into such requests as well.

2. control and sampling. I think we have a lot of open questions here
- what if there is no listener at all? I.e. tracing system. Do we expect .NET and libs to inject headers still? Does 'Trace-Context' give any guidance here?
- assuming there is a listener and it samples at least some activities. Perhaps we want to propagate context even for sampled-out activities. And assuming we still want to save some performance for sampled-out activity, the context should be propagated 'blindly' without modification. Modification (child activity) creation is not possible: imagine if Id was not hierarchical and a new child was created for sampled-out Activity. Such Activity would not be reported (as sampled out), so if some downstream service eventually would report something, if may not be possible to recover the flow - all the intermediate parent/child relationships would be lost because telemetry was not reported. 

What do you think?

Reply:
I agree with @lmolkova's suggestion to leave this out of this guide until the details are figured out. This is intended to be an entry-level guide and distributed tracing is a more advanced subject so we can briefly describe it here and point to a separate guide focused on that (once it is ready). 

Regarding:
> @TomMilos BTW, I just noticed that this guide does not cover injection/extraction at all. I think it worth mentioning.

This also is an advanced use case which might not fit well into this document. I'll try to describe it in a simple and generic way and if we are happy with it then we can leave it here. Otherwise we could direct to your Service Bus guide as a reference point for such scenario

Reply:
It is important detail. Yes, we saying that identifiers/headers should be propagated even when telemetry item was sampled out. It is crucial for multi-tenant environments with separate sampling control.

So there should be a note on the meaning of `isEnabled` and whether we will have lightweight activity population for the case when Activity is sampled, but headers still have to be propagated.